xun/functions/graph.py

```diff
@@ -21,7 +21,48 @@ def source_nodes(dag):
     return [n for n, in_degree in dag.in_degree() if in_degree == 0]
 
 
-class CallNode:
+class Unpackable:
+    def unpack(self, shape, *, _subscript=()):
+        """
+        Unpack a CallNode or CallNodeSubscript into a tuple of
+        CallNodeSubscripts with a given shape using recursion. Every
+        CallNodeSubscript holds the parent CallNode along with its respective
+        subscript index.
+
+        Examples
+        --------
+        >>> CallNode('f').unpack((2, (2,)))
+        (
+            CallNodeSubscript('f', (0,)),
+            CallNodeSubscript('f', (1,)),
+            (
+                CallNodeSubscript('f', (2, 0)),
+                CallNodeSubscript('f', (2, 1))
+            )
+        )
+        """
+        output = ()
+        idx = 0
+        for element in shape:
+            if isinstance(element, int):
+                for _ in range(element):
+                    subscript = _subscript + (idx, )
+                    idx += 1
+                    output += (CallNodeSubscript(self, subscript), )
+            elif isinstance(element, tuple):
+                subscript = _subscript + (idx, )
+                idx += 1
+                output += (self.unpack(shape=element, _subscript=subscript), )
+            elif isinstance(element, type(Ellipsis)):
+                subscript = _subscript + (idx, )
+                idx += 1
+                output += (CallNodeSubscript(self, subscript), )
+            else:
+                raise TypeError("Invalid content in shape tuple")
+        return output
```

Comment:
@JensGM is there a better way to share only this one function between the two successive classes?